stan-dev/math
# Feature/issue 848 map rect mpi, fixes #848

#### Submission Checklist


#### Summary:
Add MPI backend to map_rect which executes tasks in parallel.

#### Intended Effect:
Use multiple cores to execute independent tasks passed to map_rect.

#### How to Verify:
Please first enable MPI, see the stan-math wiki for doing so. The tests to run are

- `test/unit/math/{prim,rev}/mat/functor/map_rect_mpi_test.cpp`
- `test/unit/math/prim/mat/functor/mpi_parallel_call_test.cpp`

#### Side Effects:
Introduces the new `mpi_parallel_call` facility which comes with its own doc and tests.

#### Documentation:
Lot's of doxygen documentation.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Sebastian Weber

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)
